### PR TITLE
Desktop: Fixes #6211: Try to replace the external link with internal link when attachment file is pasted in Markdown editor

### LINKED ISSUE #6211: Joplin saves resources   with absolute Windows Paths causing broken  synced notes and .jex file is missing images
To replicate:

https://github.com/laurent22/joplin/issues/6211#issuecomment-1154139345

* * *

Hi

For some reason some images in some of the notes were not synching to other devices. I tried to figure out what the issue was but I can't pinpoint. After even more digging I found that Joplin records absolute Windows paths to the resources which ends up breaking image loading on other devices. Obviously this does not happen always and I do not know when or how Joplin decides to behave this way.

Please see the bottom of the .md content I pasted below. I renamed my user to **USER**

```

![5c72282b15413a87eb6bc17ff130689d.png](file:///C:/Users/USER/.config/joplin-desktop/resources/fad7b55125dd4500ad16c331911d0399.png)

<img src="file:///C:/Users/USER/.config/joplin-desktop/resources/1ee8144b5c6f4c32b431c869decb387c.png" alt="2fcdb0291570e35fbe09e433ad564eb3.png" width="1647" height="2687" class="jop-noMdConv">



```

.jex file ( You can see that those images are not recorded in the resources folder)

[Human head - Wikipedia 2.zip](https://github.com/laurent22/joplin/files/8150477/Human.head.-.Wikipedia.2.zip)

I exported one of the notes I found with that issue, and as you can see the resource has absolute path, this is not something I added myself. This page was captured by the Chrome extension.

This is the page I captured, but I do not know if this is really a Clipper issue since the clipper would not know where the images are saved on the drive. 

https://en.wikipedia.org/w/index.php?title=Human_head

As a result of this issue the exported .jex files also does not contain those images which has broken image icons, same with syncing to the other devices where those images are also going to be broken on those devices.

## Environment
Joplin version: 2.7.13
Platform: Windows 10

1. Select the page and capture the selection while the visual editor is enabled on Windows 
2. Sync to other devices

```
# Human head

| Human head |     |
| --- | --- |
| [![Proportions of the Head.jpg](../_resources/250px-Proportions_of_the_Head_48cce60f7c324af986fb.jpg)](https://en.wikipedia.org/wiki/File:Proportions_of_the_Head.jpg)<br>The [human](https://en.wikipedia.org/wiki/Human "Human") head drawn by [Leonardo da Vinci](https://en.wikipedia.org/wiki/Leonardo_da_Vinci "Leonardo da Vinci") |     |
| Details |     |
| Identifiers |     |
| [Latin](https://en.wikipedia.org/wiki/Latin "Latin") | *caput* |
| [MeSH](https://en.wikipedia.org/wiki/Medical_Subject_Headings "Medical Subject Headings") | [D006257](https://meshb.nlm.nih.gov/record/ui?ui=D006257) |
| [TA98](https://en.wikipedia.org/wiki/Terminologia_Anatomica "Terminologia Anatomica") | [A01.1.00.001](http://www.unifr.ch/ifaa/Public/EntryPage/TA98%20Tree/Entity%20TA98%20EN/01.1.00.001%20Entity%20TA98%20EN.htm) |
| [TA2](https://en.wikipedia.org/wiki/Terminologia_Anatomica "Terminologia Anatomica") | [98](https://ta2viewer.openanatomy.org/?id=98) |
| [FMA](https://en.wikipedia.org/wiki/Foundational_Model_of_Anatomy "Foundational Model of Anatomy") | [7154](https://bioportal.bioontology.org/ontologies/FMA/?p=classes&conceptid=http%3A%2F%2Fpurl.org%2Fsig%2Font%2Ffma%2Ffma7154) |
| ***[Anatomical terminology](https://en.wikipedia.org/wiki/Anatomical_terminology "Anatomical terminology")***<br>\[[edit on Wikidata](https://www.wikidata.org/wiki/Q3409626 "d:Q3409626")\] |     |

In human anatomy, the **head** is at the top of the [human body](https://en.wikipedia.org/wiki/Human_body "Human body"). It supports the [face](https://en.wikipedia.org/wiki/Face "Face") and is maintained by the [skull](https://en.wikipedia.org/wiki/Human_skull "Human skull"), which itself encloses the [brain](https://en.wikipedia.org/wiki/Human_brain "Human brain").

## Contents

- [1 Structure](#Structure)
    - [1.1 Blood supply](#Blood_supply)
    - [1.2 Nerve supply](#Nerve_supply)
- [2 Function](#Function)
- [3 Society and culture](#Society_and_culture)
    - [3.1 Clothing](#Clothing)
    - [3.2 Anthropometry](#Anthropometry)
        - [3.2.1 Average head sizes](#Average_head_sizes)
- [4 See also](#See_also)
- [5 References](#References)
- [6 Further reading](#Further_reading)

## [](#)Structure

[![](../_resources/220px-Head_ap_anatomy_d4e750f526004c0d8efdb42bbfca.jpg)](https://en.wikipedia.org/wiki/File:Head_ap_anatomy.jpg)

Anatomy of the human head

The human head consists of a fleshy outer portion, which surrounds the bony [skull](https://en.wikipedia.org/wiki/Human_skull "Human skull"). The [brain](https://en.wikipedia.org/wiki/Brain "Brain") is enclosed within the skull. There are 22 bones in the human head. The head rests on the [neck](https://en.wikipedia.org/wiki/Neck "Neck"), and the seven [cervical vertebrae](https://en.wikipedia.org/wiki/Cervical_vertebrae "Cervical vertebrae") support it. The human head typically weighs between 2.3 and 5 kilograms (5.1 and 11.0 lb)

The face is the [anterior](https://en.wikipedia.org/wiki/Anterior "Anterior") part of the head, containing the [eyes](https://en.wikipedia.org/wiki/Eyes "Eyes"), [nose](https://en.wikipedia.org/wiki/Nose "Nose"), and [mouth](https://en.wikipedia.org/wiki/Mouth "Mouth"). On either side of the mouth, the [cheeks](https://en.wikipedia.org/wiki/Cheek "Cheek") provide a fleshy border to the [oral cavity](https://en.wikipedia.org/wiki/Oral_cavity "Oral cavity"). The ears sit to either side of the head.

### [](#)Blood supply

The head receives blood supply through the internal and external [carotid arteries](https://en.wikipedia.org/wiki/Carotid_artery "Carotid artery"). These supply the area outside of the skull ([external carotid artery](https://en.wikipedia.org/wiki/External_carotid_artery "External carotid artery")) and inside of the skull ([internal carotid artery](https://en.wikipedia.org/wiki/Internal_carotid_artery "Internal carotid artery")). The area inside the skull also receives blood supply from the [vertebral arteries](https://en.wikipedia.org/wiki/Vertebral_artery "Vertebral artery"), which travel up through the cervical vertebrae.

### [](#)Nerve supply

[![](../_resources/350px-Gray784_313b8e9ade75475fa24b80b8d1e5977a.png)](https://en.wikipedia.org/wiki/File:Gray784.png)

Sensory areas of the head, showing the general distribution of the three divisions of the fifth nerve. From Gray’s Anatomy 1918

The twelve pairs of [cranial nerves](https://en.wikipedia.org/wiki/Cranial_nerve "Cranial nerve") provide the majority of nervous control to the head. The sensation to the face is provided by the branches of the [trigeminal nerve](https://en.wikipedia.org/wiki/Trigeminal_nerve "Trigeminal nerve"), the fifth [cranial nerve](https://en.wikipedia.org/wiki/Cranial_nerve "Cranial nerve"). Sensation to other portions of the head is provided by the [cervical nerves](https://en.wikipedia.org/wiki/Cervical_nerve "Cervical nerve").

Modern texts are in agreement about which areas of the skin are served by which nerves, but there are minor variations in some of the details. The borders designated by diagrams in the 1918 edition of [Gray’s Anatomy](https://en.wikipedia.org/wiki/Gray%27s_Anatomy "Gray's Anatomy") are similar but not identical to those generally accepted today.

The cutaneous innervation of the head is as follows:

- [Ophthalmic nerve](https://en.wikipedia.org/wiki/Ophthalmic_nerve "Ophthalmic nerve") (green)
- [Maxillary nerve](https://en.wikipedia.org/wiki/Maxillary_nerve "Maxillary nerve") (pink)
- [Mandibular nerve](https://en.wikipedia.org/wiki/Mandibular_nerve "Mandibular nerve") (yellow)
- [Cervical plexus](https://en.wikipedia.org/wiki/Cervical_plexus "Cervical plexus") (purple)
- [Dorsal rami](https://en.wikipedia.org/wiki/Dorsal_ramus_of_spinal_nerve "Dorsal ramus of spinal nerve") of [cervical nerves](https://en.wikipedia.org/wiki/Cervical_nerves "Cervical nerves") (blue) and others are in picture which show following in upper column

## [](#)Function

|     |     |
| --- | --- |
| [![\[icon\]](../_resources/20px-Wiki_letter_w_cropped.svg_e222b695cf4a4001ba7.png)](https://en.wikipedia.org/wiki/File:Wiki_letter_w_cropped.svg) | This section **needs expansion**. You can help by [adding to it](https://en.wikipedia.org/w/index.php?title=Human_head&action=edit&section=). *(March 2014)* |

The head contains [sensory organs](https://en.wikipedia.org/wiki/Sensory_organ "Sensory organ"): two eyes, two ears, a nose and [tongue](https://en.wikipedia.org/wiki/Tongue "Tongue") inside of the mouth. It also houses the [brain](https://en.wikipedia.org/wiki/Brain "Brain"). Together, these organs function as a processing center for the body by relaying sensory information to the brain. Humans can process information faster by having this central nerve cluster.

## [](#)Society and culture

For [humans](https://en.wikipedia.org/wiki/Human "Human"), the front of the head (the face) is the main distinguishing feature between different people due to its easily discernible features, such as eye and hair colors, shapes of the sensory organs, and the wrinkles. Humans easily differentiate between faces because of the brain’s predisposition toward facial recognition. When observing a relatively unfamiliar species, all faces seem nearly identical. Human infants are biologically programmed to recognize subtle differences in anthropomorphic facial features.<sup>[\[1\]](#cite_note-1)</sup>

[![](../_resources/220px-COLLECTIE_TROPENMUSEUM_Een_510111e15f12488c9.jpg)](https://en.wikipedia.org/wiki/File:COLLECTIE_TROPENMUSEUM_Een_Ibu_Dajak_krijger_uit_Long_Nawan_Z._en_O._afdeling_Borneo._TMnr_60034031.jpg)

[Dayak people](https://en.wikipedia.org/wiki/Dayak_people "Dayak people") were feared for their headhunting practices

People who have greater than average intelligence are sometimes depicted in cartoons as having bigger heads as a way of notionally indicating that they have a “larger brain”. Additionally, in [science fiction](https://en.wikipedia.org/wiki/Science_fiction "Science fiction"), an [extraterrestrial](https://en.wikipedia.org/wiki/Extraterrestrial_life_in_popular_culture "Extraterrestrial life in popular culture") having a big head is often symbolic of high intelligence. Despite this depiction, advances in [neurobiology](https://en.wikipedia.org/wiki/Neurobiology "Neurobiology") have shown that the functional diversity of the brain means that a difference in overall brain size is only slightly to moderately correlated to differences in overall intelligence between two humans.<sup>[\[2\]](#cite_note-2)</sup>

The head is a source for many metaphors and [metonymies](https://en.wikipedia.org/wiki/Metonymy "Metonymy") in [human language](https://en.wikipedia.org/wiki/Human_language "Human language"), including referring to things typically near the human head ( “the head of the bed”), things physically similar to the way a head is arranged spatially to a body (“the head of the table”), metaphorically (“the head of the class”), and things that represent some characteristics associated with the head, such as intelligence (“there are a lot of good heads in this company”).<sup>[\[3\]](#cite_note-3)</sup>

Ancient Greeks had a method for evaluating [sexual attractiveness](https://en.wikipedia.org/wiki/Sexual_attractiveness "Sexual attractiveness") based on the [Golden ratio](https://en.wikipedia.org/wiki/Golden_ratio "Golden ratio"), part of which included measurements of the head.<sup>[\[4\]](#cite_note-4)</sup>

[Headhunting](https://en.wikipedia.org/wiki/Headhunting "Headhunting") is the practice of taking and [preserving a person’s head](https://en.wikipedia.org/wiki/Human_trophy_collecting "Human trophy collecting") after killing the person. Headhunting has been practiced across the Americas, Europe, Asia, and Oceania for millennia.<sup>[\[5\]](#cite_note-5)</sup>

### [](#)Clothing

[![](../_resources/170px-0004_Madagascar_(555870032_d82430257b3b43b8a.jpg)](https://en.wikipedia.org/wiki/File:0004_Madagascar_%285558700320%29.jpg)

A man wearing a [straw hat](https://en.wikipedia.org/wiki/Straw_hat "Straw hat")

Main article: [Headgear](https://en.wikipedia.org/wiki/Headgear "Headgear")

Headpieces can signify status, origin, religious/spiritual beliefs, social grouping, team affiliation, occupation, or fashion choices.

In many cultures, covering the head is seen as a sign of respect. Often, some or all of the head must be covered and veiled when entering holy places or places of prayer. For many centuries, women in Europe, the [Middle East](https://en.wikipedia.org/wiki/Middle_East "Middle East"), and [South Asia](https://en.wikipedia.org/wiki/South_Asia "South Asia") have covered their head hair as a sign of modesty. This trend has changed drastically in Europe in the 20th century, although is still observed in other parts of the world. In addition, a number of religions require men to wear specific head clothing—such as the [Islamic](https://en.wikipedia.org/wiki/Islamic "Islamic") [Taqiyah (cap)](https://en.wikipedia.org/wiki/Taqiyah_%28cap%29 "Taqiyah (cap)"), [Jewish](https://en.wikipedia.org/wiki/Jewish "Jewish") [yarmulke](https://en.wikipedia.org/wiki/Yarmulke "Yarmulke"), or the [Sikh turban](https://en.wikipedia.org/wiki/Sikh_turban "Sikh turban"). The same goes for women with the [Muslim](https://en.wikipedia.org/wiki/Muslim "Muslim") [hijab](https://en.wikipedia.org/wiki/Hijab "Hijab") or [Christian](https://en.wikipedia.org/wiki/Christians "Christians") nun’s [habit](https://en.wikipedia.org/wiki/Religious_habit "Religious habit").

A [hat](https://en.wikipedia.org/wiki/Hat "Hat") is a head covering that can serve a variety of purposes. Hats may be worn as part of a uniform or used as a protective device, such as a [hard hat](https://en.wikipedia.org/wiki/Hard_hat "Hard hat"), a covering for warmth, or a fashion accessory. Hats can also be indicative of social status in some areas of the world.

### [](#)[Anthropometry](https://en.wikipedia.org/wiki/Anthropometry "Anthropometry")

While numerous charts detailing head sizes in infants and children exist, most do not measure average head circumference past the age of 21. Reference charts for adult head circumference also generally feature [homogeneous](https://en.wikipedia.org/wiki/Homogeneous "Homogeneous") samples and fail to take height and weight into account.<sup>[\[6\]](#cite_note-6)</sup>

One study in the United States estimated the average human head circumference to be 57 centimetres (22+1⁄2 in) in males and 55 centimetres (21+3⁄4 in) in females.<sup>[\[7\]](#cite_note-7)</sup><sup>\[*[dubious](https://en.wikipedia.org/wiki/Wikipedia:Accuracy_dispute#Disputed_statement "Wikipedia:Accuracy dispute") – [discuss](https://en.wikipedia.org/wiki/Talk:Human_head#Talk_Average_Head_Circumference_Statement "Talk:Human head")*\]</sup> A British study by [Newcastle University](https://en.wikipedia.org/wiki/Newcastle_University "Newcastle University") showed an average size of 57.2 cm for males and 55.2 cm for females with average size varying proportionally with height <sup>[\[8\]](#cite_note-8)</sup>

[Macrocephaly](https://en.wikipedia.org/wiki/Macrocephaly "Macrocephaly") can be an indicator of increased risk for some types of cancer in individuals who carry the genetic mutation that causes [Cowden syndrome](https://en.wikipedia.org/wiki/Cowden_syndrome "Cowden syndrome"). For adults, this refers to head sizes greater than 58 centimeters in men or greater than 57 centimeters in women.<sup>[\[9\]](#cite_note-9)</sup><sup>[\[10\]](#cite_note-10)</sup>

#### [](#)Average head sizes

![5c72282b15413a87eb6bc17ff130689d.png](file:///C:/Users/USER/.config/joplin-desktop/resources/fad7b55125dd4500ad16c331911d0399.png)

<img src="file:///C:/Users/USER/.config/joplin-desktop/resources/1ee8144b5c6f4c32b431c869decb387c.png" alt="2fcdb0291570e35fbe09e433ad564eb3.png" width="1647" height="2687" class="jop-noMdConv">

```

### PR DESCRIPTION
**Pull request by wh201906**
This will fix #6211. When the user copies an image from WYSIWYG editor, the clipboard will hold the absolute path of the image. The bug is when pasting it to Markdown editor, the absolute path will be pasted. Codes in this PR will try to replace the absolute path with the internal link if possible, just like copying and pasting an image in WYSIWYG editor.
I tested it on Ubuntu and the bug is fixed.
I have no idea about how to call an async function in an sync function. I just imitated the existing code. Please tell me if my code is wrong or not appropriate and I will fix it.

### CODE REVIEW

**Comment by wh201906** [comment]
~~I think editorPasteText() should run `editorRef.current.replaceSelection(clipboard.readText())` if the `replaceResourceExternalToInternalLinks()` returns an empty string or "promise gets rejected". But I don't know how to implement this.~~
The async function `replaceResourceExternalToInternalLinks(body, options)` just replace the specified text in body, so there is no need to add extra code to handle the empty result. Also, this function doesn't throw errors, and it doesn't call other async functions, so the promise is always resolved.

**Comment by wh201906** [comment]
I don't know why the mobile build in CI failed. I did build the desktop version on Ubuntu and it works.

Here are the logs for `Main (ubuntu-latest)`
```
2022-09-19T03:08:29.2443537Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m [03:08:29] Starting 'encodeAssets'...
2022-09-19T03:08:29.2517417Z [94m➤[39m [90mYN0000[39m: │ [38;5;166m@joplin/[39m[38;5;173mapp-mobile[39m[38;5;111m@[39m[38;5;111mworkspace:packages/app-mobile[39m [31mSTDERR[39m [03:08:29] 'encodeAssets' errored after 172 ms
2022-09-19T03:08:29.2524888Z [94m➤[39m [90mYN0000[39m: │ [38;5;166m@joplin/[39m[38;5;173mapp-mobile[39m[38;5;111m@[39m[38;5;111mworkspace:packages/app-mobile[39m [31mSTDERR[39m [03:08:29] Error: ENOENT: no such file or directory, open '/home/runner/work/joplin/joplin/packages/app-mobile/tools/../pluginAssets/katex/fonts/KaTeX_SansSerif-Regular.woff2.base64.js'
2022-09-19T03:08:29.2526917Z [94m➤[39m [90mYN0000[39m: │ [38;5;166m@joplin/[39m[38;5;173mapp-mobile[39m[38;5;111m@[39m[38;5;111mworkspace:packages/app-mobile[39m [31mSTDERR[39m [03:08:29] 'build' errored after 59 s
2022-09-19T03:08:29.3598695Z [91m➤[39m YN0009: │ [38;5;166m@joplin/[39m[38;5;173mapp-mobile[39m[38;5;111m@[39m[38;5;111mworkspace:packages/app-mobile[39m couldn't be built successfully (exit code 1, logs can be found here: [38;5;170m/tmp/xfs-ecaa907d/build.log[39m)
2022-09-19T03:08:29.7358828Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m [03:08:29] Finished 'encodeAssets' after 491 ms
```
Here are the logs for `ServerDockerImage (ubuntu-latest)`
```
2022-09-19T03:07:12.8879815Z [94m➤[39m [90mYN0000[39m: │ [38;5;166m@joplin/[39m[38;5;173mapp-mobile[39m[38;5;111m@[39m[38;5;111mworkspace:packages/app-mobile[39m [32mSTDOUT[39m [03:07:12] Starting 'encodeAssets'...
2022-09-19T03:07:12.9020328Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m [03:07:12] 'encodeAssets' errored after 122 ms
2022-09-19T03:07:12.9029337Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m [03:07:12] Error: EEXIST: file already exists, mkdir '/home/runner/work/joplin/joplin/packages/app-mobile/pluginAssets/katex/fonts'
2022-09-19T03:07:12.9040022Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m [03:07:12] 'build' errored after 55 s
2022-09-19T03:07:12.9816960Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [94m➤[39m [90mYN0000[39m: [38;5;214m[@joplin/app-mobile]:[39m Process exited (exit code 1), completed in 1m 2s
2022-09-19T03:07:12.9822495Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [91m➤[39m YN0000: The command failed for workspaces that are depended upon by other workspaces; can't satisfy the dependency graph
2022-09-19T03:07:12.9829183Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [32mSTDOUT[39m [91m➤[39m YN0000: Failed with errors in 1m 4s
2022-09-19T03:07:13.1364868Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [31mSTDERR[39m [03:07:13] 'build' errored after 1.12 min
2022-09-19T03:07:13.1365889Z [94m➤[39m [90mYN0000[39m: │ [38;5;173mroot[39m[38;5;111m@[39m[38;5;111mworkspace:.[39m [31mSTDERR[39m [03:07:13] Error: Command failed with exit code 1: yarn run buildSequential
```
They all have the similar errors when doing `encodeAssets` in the building process of `app-mobile`, but I didn't touch the code in `app-mobile`. Maybe we need to re-run the jobs

**Comment by laurent22** [comment]
We need a clear description of what's been fixed, and a working pr.

**Comment by wh201906** [comment]
> We need a clear description of what's been fixed, and a working pr.

I have updated the title and description of this PR. Are they correct?
I think the PR is working now since it passed all CI test, and I have tested the functionality of it on Ubuntu and it works as expected.

File: packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx [hunk 1]
```diff
@@ -293,7 +294,10 @@ function CodeMirror(props: NoteBodyEditorProps, ref: any) {
 
 	const editorPasteText = useCallback(() => {
 		if (editorRef.current) {
-			editorRef.current.replaceSelection(clipboard.readText());
+			void Note.replaceResourceExternalToInternalLinks(clipboard.readText(), { useAbsolutePaths: true })
+				.then((modifiedMd) => {
+					editorRef.current.replaceSelection(modifiedMd);
+				});
```
Comment (laurent22) [inline packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx]: Don't use .then(), use async/await

Comment (wh201906) [reply packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/CodeMirror.tsx]: ~~I don't know if my code in [d9d8f60](https://github.com/laurent22/joplin/pull/6865/commits/d9d8f60e2a67f984627fea02a400ea3d35930532) is right.~~ It passed the CI test, then I tested it on Ubuntu and it works as expected.

**Comment by wh201906** [comment]
@laurent22 Hi. Is there anything I need to do?

**Comment by laurent22** [comment]
Thanks for fixing this, and I think that's the right fix, but I wonder if we want to support this kind of copy and paste between wysiwyg and markdown editor? Lots of other things probably don't work, such as copying tables, and it feels like the solution is simply to not do it.

Why would someone want to copy from one editor to another anyway? Why not stay on the same one?

**Comment by wh201906** [comment]
> but I wonder if we want to support this kind of copy and paste between wysiwyg and markdown editor?

I guess this PR can be applied to more situations. Any external links referring to something in profile(resources) directory can be converted into the internal links.

> Lots of other things probably don't work, such as copying tables

I don't know why these things won't work after merging this PR. Could you please give me more information or examples? 
Plus, this PR is based on the behaviors of TinyMCE editor, as I mentioned in [this comment](https://github.com/laurent22/joplin/issues/6211#issuecomment-1250337502). If something doesn't work there, the TinyMCE editor should have the similar problem.

**Comment by wh201906** [comment]
Another solution is to find the resource parts in the clipboard which look like `[...](...)`, then replace them only.

**Comment by tomasz1986** [comment]
> Why would someone want to copy from one editor to another anyway? Why not stay on the same one?

I mainly use WYSIWYG, but I also switch between the two when I want to add formatting that isn't available in the top bar (e.g. H4/5/6 or straight-up HTML code), and if I happen to have an image that I want to insert copied to clipboard already, I may as well paste it in the Markdown editor as well.

The main problem here is probably that the non-technical user isn't really aware that such a link is absolute and device specific. With tables and other complex formatting, the user can immediately see when something isn't right. With these images, it's only after they've synced their data to another machine when they discover that the links are broken. In the worst case scenario, at that point they may not even have the original images anymore.

**Comment by wh201906** [comment]
I think this problem is easier to be triggered than we expected. 
https://discourse.joplinapp.org/t/why-is-github-issues-long-left-to-be-answered/25993
https://discourse.joplinapp.org/t/disappearing-accessories-a-serious-functional-defect/28082
I'm willing to renovate this PR if the current version would cause any problem.

**Comment by tomasz1986** [comment]
> I think this problem is easier to be triggered than we expected.

Yeah, and I think another aspect which hasn't been pointed enough yet is that this issue literally causes the user to lose data (e.g. when they no longer have access to the original device and/or file). The user who has lost data due to a bug is usually a very unhappy user.

**Comment by laurent22** [comment]
In order to contribute to the project, it is now required to sign a Contributor License Agreement (CLA) as described in this announcement: [Joplin is switching to the GNU Affero General Public License v3 (AGPL-3.0)](https://joplinapp.org/news/20221221-agpl/). Your contributions will remain open source and you only need to sign this document once. Please download it from [here](https://github.com/laurent22/joplin/blob/dev/readme/cla.md) and send it over. If you have any question about it please let me know.

**Comment by laurent22** [comment]
recheck

**Comment by wh201906** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by laurent22** [comment]
Ok let's merge, this change seems relatively safe and still an improvement. Thanks for implementing this @wh201906!

**Pull request merged**